// @certusone/wormhole-sdk/lib/cjs/token_bridge/redeem.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.redeemOnAptos = exports.redeemOnNear = exports.redeemOnAlgorand = exports.redeemOnSolana = exports.redeemAndUnwrapOnSolana = exports.redeemOnXpla = exports.redeemOnInjective = exports.submitVAAOnInjective = exports.redeemOnTerra = exports.redeemOnEthNative = exports.redeemOnEth = void 0;
var spl_token_1 = require("@solana/spl-token");
var web3_js_1 = require("@solana/web3.js");
var terra_js_1 = require("@terra-money/terra.js");
var js_base64_1 = require("js-base64");
var bn_js_1 = __importDefault(require("bn.js"));
var algorand_1 = require("../algorand");
var ethers_contracts_1 = require("../ethers-contracts");
var utils_1 = require("../utils");
var sdk_ts_1 = require("@injectivelabs/sdk-ts");
var tokenBridge_1 = require("../solana/tokenBridge");
var vaa_1 = require("../vaa");
var getForeignAsset_1 = require("./getForeignAsset");
var xpla_js_1 = require("@xpla/xpla.js");
var aptos_1 = require("../aptos");
function redeemOnEth(tokenBridgeAddress, signer, signedVAA, overrides) {
    if (overrides === void 0) { overrides = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var bridge, v, receipt;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    bridge = ethers_contracts_1.Bridge__factory.connect(tokenBridgeAddress, signer);
                    return [4 /*yield*/, bridge.completeTransfer(signedVAA, overrides)];
                case 1:
                    v = _a.sent();
                    return [4 /*yield*/, v.wait()];
                case 2:
                    receipt = _a.sent();
                    return [2 /*return*/, receipt];
            }
        });
    });
}
exports.redeemOnEth = redeemOnEth;
function redeemOnEthNative(tokenBridgeAddress, signer, signedVAA, overrides) {
    if (overrides === void 0) { overrides = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var bridge, v, receipt;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    bridge = ethers_contracts_1.Bridge__factory.connect(tokenBridgeAddress, signer);
                    return [4 /*yield*/, bridge.completeTransferAndUnwrapETH(signedVAA, overrides)];
                case 1:
                    v = _a.sent();
                    return [4 /*yield*/, v.wait()];
                case 2:
                    receipt = _a.sent();
                    return [2 /*return*/, receipt];
            }
        });
    });
}
exports.redeemOnEthNative = redeemOnEthNative;
function redeemOnTerra(tokenBridgeAddress, walletAddress, signedVAA) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new terra_js_1.MsgExecuteContract(walletAddress, tokenBridgeAddress, {
                    submit_vaa: {
                        data: js_base64_1.fromUint8Array(signedVAA),
                    },
                })];
        });
    });
}
exports.redeemOnTerra = redeemOnTerra;
/**
 * Submits the supplied VAA to Injective
 * @param tokenBridgeAddress Address of Inj token bridge contract
 * @param walletAddress Address of wallet in inj format
 * @param signedVAA VAA with the attestation message
 * @returns Message to be broadcast
 */
function submitVAAOnInjective(tokenBridgeAddress, walletAddress, signedVAA) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, sdk_ts_1.MsgExecuteContractCompat.fromJSON({
                    contractAddress: tokenBridgeAddress,
                    sender: walletAddress,
                    exec: {
                        msg: {
                            data: js_base64_1.fromUint8Array(signedVAA),
                        },
                        action: "submit_vaa",
                    },
                })];
        });
    });
}
exports.submitVAAOnInjective = submitVAAOnInjective;
exports.redeemOnInjective = submitVAAOnInjective;
function redeemOnXpla(tokenBridgeAddress, walletAddress, signedVAA) {
    return new xpla_js_1.MsgExecuteContract(walletAddress, tokenBridgeAddress, {
        submit_vaa: {
            data: js_base64_1.fromUint8Array(signedVAA),
        },
    });
}
exports.redeemOnXpla = redeemOnXpla;
function redeemAndUnwrapOnSolana(connection, bridgeAddress, tokenBridgeAddress, payerAddress, signedVaa, commitment) {
    return __awaiter(this, void 0, void 0, function () {
        var parsed, targetPublicKey, targetAmount, rentBalance, payerPublicKey, ancillaryKeypair, completeTransferIx, createAncillaryAccountIx, initAccountIx, balanceTransferIx, closeAccountIx, blockhash, transaction;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    parsed = vaa_1.parseTokenTransferVaa(signedVaa);
                    targetPublicKey = new web3_js_1.PublicKey(parsed.to);
                    return [4 /*yield*/, spl_token_1.getMint(connection, spl_token_1.NATIVE_MINT, commitment).then(function (info) {
                            return parsed.amount * BigInt(Math.pow(10, info.decimals - utils_1.MAX_VAA_DECIMALS));
                        })];
                case 1:
                    targetAmount = _a.sent();
                    return [4 /*yield*/, spl_token_1.getMinimumBalanceForRentExemptAccount(connection, commitment)];
                case 2:
                    rentBalance = _a.sent();
                    if (Buffer.compare(parsed.tokenAddress, spl_token_1.NATIVE_MINT.toBuffer()) != 0) {
                        return [2 /*return*/, Promise.reject("tokenAddress != NATIVE_MINT")];
                    }
                    payerPublicKey = new web3_js_1.PublicKey(payerAddress);
                    ancillaryKeypair = web3_js_1.Keypair.generate();
                    completeTransferIx = tokenBridge_1.createCompleteTransferNativeInstruction(tokenBridgeAddress, bridgeAddress, payerPublicKey, signedVaa);
                    createAncillaryAccountIx = web3_js_1.SystemProgram.createAccount({
                        fromPubkey: payerPublicKey,
                        newAccountPubkey: ancillaryKeypair.publicKey,
                        lamports: rentBalance,
                        space: spl_token_1.ACCOUNT_SIZE,
                        programId: spl_token_1.TOKEN_PROGRAM_ID,
                    });
                    initAccountIx = spl_token_1.createInitializeAccountInstruction(ancillaryKeypair.publicKey, spl_token_1.NATIVE_MINT, payerPublicKey);
                    balanceTransferIx = spl_token_1.createTransferInstruction(targetPublicKey, ancillaryKeypair.publicKey, payerPublicKey, targetAmount.valueOf());
                    closeAccountIx = spl_token_1.createCloseAccountInstruction(ancillaryKeypair.publicKey, //account to close
                    payerPublicKey, //Remaining funds destination
                    payerPublicKey //authority
                    );
                    return [4 /*yield*/, connection.getLatestBlockhash(commitment)];
                case 3:
                    blockhash = (_a.sent()).blockhash;
                    transaction = new web3_js_1.Transaction();
                    transaction.recentBlockhash = blockhash;
                    transaction.feePayer = payerPublicKey;
                    transaction.add(completeTransferIx, createAncillaryAccountIx, initAccountIx, balanceTransferIx, closeAccountIx);
                    transaction.partialSign(ancillaryKeypair);
                    return [2 /*return*/, transaction];
            }
        });
    });
}
exports.redeemAndUnwrapOnSolana = redeemAndUnwrapOnSolana;
function redeemOnSolana(connection, bridgeAddress, tokenBridgeAddress, payerAddress, signedVaa, feeRecipientAddress, commitment) {
    return __awaiter(this, void 0, void 0, function () {
        var parsed, createCompleteTransferInstruction, transaction, blockhash;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    parsed = vaa_1.parseTokenTransferVaa(signedVaa);
                    createCompleteTransferInstruction = parsed.tokenChain == utils_1.CHAIN_ID_SOLANA
                        ? tokenBridge_1.createCompleteTransferNativeInstruction
                        : tokenBridge_1.createCompleteTransferWrappedInstruction;
                    transaction = new web3_js_1.Transaction().add(createCompleteTransferInstruction(tokenBridgeAddress, bridgeAddress, payerAddress, parsed, feeRecipientAddress));
                    return [4 /*yield*/, connection.getLatestBlockhash(commitment)];
                case 1:
                    blockhash = (_a.sent()).blockhash;
                    transaction.recentBlockhash = blockhash;
                    transaction.feePayer = new web3_js_1.PublicKey(payerAddress);
                    return [2 /*return*/, transaction];
            }
        });
    });
}
exports.redeemOnSolana = redeemOnSolana;
/**
 * This basically just submits the VAA to Algorand
 * @param client AlgodV2 client
 * @param tokenBridgeId Token bridge ID
 * @param bridgeId Core bridge ID
 * @param vaa The VAA to be redeemed
 * @param acct Sending account
 * @returns Transaction ID(s)
 */
function redeemOnAlgorand(client, tokenBridgeId, bridgeId, vaa, senderAddr) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, algorand_1._submitVAAAlgorand(client, tokenBridgeId, bridgeId, vaa, senderAddr)];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.redeemOnAlgorand = redeemOnAlgorand;
function redeemOnNear(provider, account, tokenBridge, vaa) {
    return __awaiter(this, void 0, void 0, function () {
        var options, p, _a, found, receiver, token, bal, bal_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    options = [];
                    p = algorand_1._parseVAAAlgorand(vaa);
                    if (p.ToChain !== utils_1.CHAIN_ID_NEAR) {
                        throw new Error("Not destined for NEAR");
                    }
                    return [4 /*yield*/, utils_1.hashLookup(provider, tokenBridge, utils_1.uint8ArrayToHex(p.ToAddress))];
                case 1:
                    _a = _b.sent(), found = _a.found, receiver = _a.value;
                    if (!found) {
                        throw new Error("Unregistered receiver (receiving account is not registered)");
                    }
                    return [4 /*yield*/, getForeignAsset_1.getForeignAssetNear(provider, tokenBridge, p.FromChain, p.Contract)];
                case 2:
                    token = _b.sent();
                    if (!(p.Contract !==
                        "0000000000000000000000000000000000000000000000000000000000000000")) return [3 /*break*/, 5];
                    if (token === "" || token === null) {
                        throw new Error("Unregistered token (has it been attested?)");
                    }
                    return [4 /*yield*/, utils_1.callFunctionNear(provider, token, "storage_balance_of", {
                            account_id: receiver,
                        })];
                case 3:
                    bal = _b.sent();
                    if (bal === null) {
                        options.push({
                            contractId: token,
                            methodName: "storage_deposit",
                            args: { account_id: receiver, registration_only: true },
                            gas: new bn_js_1.default("100000000000000"),
                            attachedDeposit: new bn_js_1.default("2000000000000000000000"), // 0.002 NEAR
                        });
                    }
                    if (!(p.Fee !== undefined &&
                        Buffer.compare(p.Fee, Buffer.from("0000000000000000000000000000000000000000000000000000000000000000", "hex")) !== 0)) return [3 /*break*/, 5];
                    return [4 /*yield*/, utils_1.callFunctionNear(provider, token, "storage_balance_of", {
                            account_id: account,
                        })];
                case 4:
                    bal_1 = _b.sent();
                    if (bal_1 === null) {
                        options.push({
                            contractId: token,
                            methodName: "storage_deposit",
                            args: { account_id: account, registration_only: true },
                            gas: new bn_js_1.default("100000000000000"),
                            attachedDeposit: new bn_js_1.default("2000000000000000000000"), // 0.002 NEAR
                        });
                    }
                    _b.label = 5;
                case 5:
                    options.push({
                        contractId: tokenBridge,
                        methodName: "submit_vaa",
                        args: {
                            vaa: utils_1.uint8ArrayToHex(vaa),
                        },
                        attachedDeposit: new bn_js_1.default("100000000000000000000000"),
                        gas: new bn_js_1.default("150000000000000"),
                    });
                    options.push({
                        contractId: tokenBridge,
                        methodName: "submit_vaa",
                        args: {
                            vaa: utils_1.uint8ArrayToHex(vaa),
                        },
                        attachedDeposit: new bn_js_1.default("100000000000000000000000"),
                        gas: new bn_js_1.default("150000000000000"),
                    });
                    return [2 /*return*/, options];
            }
        });
    });
}
exports.redeemOnNear = redeemOnNear;
/**
 * Register the token specified in the given VAA in the transfer recipient's account if necessary
 * and complete the transfer.
 * @param client Client used to transfer data to/from Aptos node
 * @param tokenBridgeAddress Address of token bridge
 * @param transferVAA Bytes of transfer VAA
 * @returns Transaction payload
 */
function redeemOnAptos(client, tokenBridgeAddress, transferVAA) {
    return aptos_1.completeTransferAndRegister(client, tokenBridgeAddress, transferVAA);
}
exports.redeemOnAptos = redeemOnAptos;
