// @certusone/wormhole-sdk/lib/esm/solana/nftBridge/instructions/completeNative.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
import { PublicKey, SystemProgram, SYSVAR_RENT_PUBKEY, } from "@solana/web3.js";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { createReadOnlyNftBridgeProgramInterface, tokenIdToMint, } from "../program";
import { deriveClaimKey, derivePostedVaaKey } from "../../wormhole";
import { deriveEndpointKey, deriveNftBridgeConfigKey, deriveCustodyKey, deriveCustodySignerKey, } from "../accounts";
import { isBytes, parseNftTransferVaa, } from "../../../vaa";
export function createCompleteTransferNativeInstruction(nftBridgeProgramId, wormholeProgramId, payer, vaa, toAuthority) {
    var methods = createReadOnlyNftBridgeProgramInterface(nftBridgeProgramId).methods.completeNative();
    // @ts-ignore
    return methods._ixFn.apply(methods, __spreadArray(__spreadArray([], __read(methods._args)), [{
            accounts: getCompleteTransferNativeAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa, toAuthority),
            signers: undefined,
            remainingAccounts: undefined,
            preInstructions: undefined,
            postInstructions: undefined,
        }]));
}
export function getCompleteTransferNativeAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa, toAuthority) {
    var parsed = isBytes(vaa) ? parseNftTransferVaa(vaa) : vaa;
    // the mint key is encoded in the tokenId when it was transferred out
    var mint = tokenIdToMint(parsed.tokenId);
    return {
        payer: new PublicKey(payer),
        config: deriveNftBridgeConfigKey(nftBridgeProgramId),
        vaa: derivePostedVaaKey(wormholeProgramId, parsed.hash),
        claim: deriveClaimKey(nftBridgeProgramId, parsed.emitterAddress, parsed.emitterChain, parsed.sequence),
        endpoint: deriveEndpointKey(nftBridgeProgramId, parsed.emitterChain, parsed.emitterAddress),
        to: new PublicKey(parsed.to),
        toAuthority: new PublicKey(toAuthority === undefined ? payer : toAuthority),
        custody: deriveCustodyKey(nftBridgeProgramId, mint),
        mint: mint,
        custodySigner: deriveCustodySignerKey(nftBridgeProgramId),
        rent: SYSVAR_RENT_PUBKEY,
        systemProgram: SystemProgram.programId,
        tokenProgram: TOKEN_PROGRAM_ID,
        wormholeProgram: new PublicKey(wormholeProgramId),
    };
}
