// @certusone/wormhole-sdk/lib/esm/solana/nftBridge/instructions/governance.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
import { PublicKey, SystemProgram, SYSVAR_CLOCK_PUBKEY, SYSVAR_RENT_PUBKEY, } from "@solana/web3.js";
import { createReadOnlyNftBridgeProgramInterface } from "../program";
import { deriveClaimKey, derivePostedVaaKey } from "../../wormhole";
import { deriveEndpointKey, deriveNftBridgeConfigKey, deriveUpgradeAuthorityKey, } from "../accounts";
import { isBytes, parseNftBridgeRegisterChainVaa, parseNftBridgeUpgradeContractVaa, } from "../../../vaa";
import { BpfLoaderUpgradeable, deriveUpgradeableProgramKey } from "../../utils";
export function createRegisterChainInstruction(nftBridgeProgramId, wormholeProgramId, payer, vaa) {
    var methods = createReadOnlyNftBridgeProgramInterface(nftBridgeProgramId).methods.registerChain();
    // @ts-ignore
    return methods._ixFn.apply(methods, __spreadArray(__spreadArray([], __read(methods._args)), [{
            accounts: getRegisterChainAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa),
            signers: undefined,
            remainingAccounts: undefined,
            preInstructions: undefined,
            postInstructions: undefined,
        }]));
}
export function getRegisterChainAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa) {
    var parsed = isBytes(vaa) ? parseNftBridgeRegisterChainVaa(vaa) : vaa;
    return {
        payer: new PublicKey(payer),
        config: deriveNftBridgeConfigKey(nftBridgeProgramId),
        endpoint: deriveEndpointKey(nftBridgeProgramId, parsed.foreignChain, parsed.foreignAddress),
        vaa: derivePostedVaaKey(wormholeProgramId, parsed.hash),
        claim: deriveClaimKey(nftBridgeProgramId, parsed.emitterAddress, parsed.emitterChain, parsed.sequence),
        rent: SYSVAR_RENT_PUBKEY,
        systemProgram: SystemProgram.programId,
        wormholeProgram: new PublicKey(wormholeProgramId),
    };
}
export function createUpgradeContractInstruction(nftBridgeProgramId, wormholeProgramId, payer, vaa, spill) {
    var methods = createReadOnlyNftBridgeProgramInterface(nftBridgeProgramId).methods.upgradeContract();
    // @ts-ignore
    return methods._ixFn.apply(methods, __spreadArray(__spreadArray([], __read(methods._args)), [{
            accounts: getUpgradeContractAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa, spill),
            signers: undefined,
            remainingAccounts: undefined,
            preInstructions: undefined,
            postInstructions: undefined,
        }]));
}
export function getUpgradeContractAccounts(nftBridgeProgramId, wormholeProgramId, payer, vaa, spill) {
    var parsed = isBytes(vaa) ? parseNftBridgeUpgradeContractVaa(vaa) : vaa;
    return {
        payer: new PublicKey(payer),
        vaa: derivePostedVaaKey(wormholeProgramId, parsed.hash),
        claim: deriveClaimKey(nftBridgeProgramId, parsed.emitterAddress, parsed.emitterChain, parsed.sequence),
        upgradeAuthority: deriveUpgradeAuthorityKey(nftBridgeProgramId),
        spill: new PublicKey(spill === undefined ? payer : spill),
        implementation: new PublicKey(parsed.newContract),
        programData: deriveUpgradeableProgramKey(nftBridgeProgramId),
        nftBridgeProgram: new PublicKey(nftBridgeProgramId),
        rent: SYSVAR_RENT_PUBKEY,
        clock: SYSVAR_CLOCK_PUBKEY,
        bpfLoaderUpgradeable: BpfLoaderUpgradeable.programId,
        systemProgram: SystemProgram.programId,
    };
}
