// @acala-network/contracts/typechain/factories/contracts/honzon/IHonzon__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHonzon__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                indexed: false,
                internalType: "int128",
                name: "collateralAdjustment",
                type: "int128",
            },
            {
                indexed: false,
                internalType: "int128",
                name: "debitAdjustment",
                type: "int128",
            },
        ],
        name: "AdjustedLoan",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
        ],
        name: "ClosedLoanByDex",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                internalType: "int128",
                name: "collateralAdjustment",
                type: "int128",
            },
            {
                internalType: "int128",
                name: "debitAdjustment",
                type: "int128",
            },
        ],
        name: "adjustLoan",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "maxCollateralAmount",
                type: "uint256",
            },
        ],
        name: "closeLoanByDex",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
        ],
        name: "getCollateralParameters",
        outputs: [
            {
                internalType: "uint256[]",
                name: "",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "who",
                type: "address",
            },
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
        ],
        name: "getCurrentCollateralRatio",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
        ],
        name: "getDebitExchangeRate",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "who",
                type: "address",
            },
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
        ],
        name: "getPosition",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IHonzon__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IHonzon__factory = IHonzon__factory;
IHonzon__factory.abi = _abi;
