// @certusone/wormhole-sdk/lib/cjs/solana/nftBridge/coder/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftBridgeCoder = exports.NftBridgeInstruction = void 0;
var accounts_1 = require("./accounts");
var events_1 = require("./events");
var instruction_1 = require("./instruction");
var state_1 = require("./state");
var types_1 = require("./types");
var instruction_2 = require("./instruction");
Object.defineProperty(exports, "NftBridgeInstruction", { enumerable: true, get: function () { return instruction_2.NftBridgeInstruction; } });
var NftBridgeCoder = /** @class */ (function () {
    function NftBridgeCoder(idl) {
        this.instruction = new instruction_1.NftBridgeInstructionCoder(idl);
        this.accounts = new accounts_1.NftBridgeAccountsCoder(idl);
        this.state = new state_1.NftBridgeStateCoder(idl);
        this.events = new events_1.NftBridgeEventsCoder(idl);
        this.types = new types_1.NftBridgeTypesCoder(idl);
    }
    return NftBridgeCoder;
}());
exports.NftBridgeCoder = NftBridgeCoder;
