// @certusone/wormhole-sdk/lib/cjs/solana/tokenBridge/coder/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenBridgeCoder = exports.TokenBridgeInstruction = void 0;
var accounts_1 = require("./accounts");
var events_1 = require("./events");
var instruction_1 = require("./instruction");
var state_1 = require("./state");
var types_1 = require("./types");
var instruction_2 = require("./instruction");
Object.defineProperty(exports, "TokenBridgeInstruction", { enumerable: true, get: function () { return instruction_2.TokenBridgeInstruction; } });
var TokenBridgeCoder = /** @class */ (function () {
    function TokenBridgeCoder(idl) {
        this.instruction = new instruction_1.TokenBridgeInstructionCoder(idl);
        this.accounts = new accounts_1.TokenBridgeAccountsCoder(idl);
        this.state = new state_1.TokenBridgeStateCoder(idl);
        this.events = new events_1.TokenBridgeEventsCoder(idl);
        this.types = new types_1.TokenBridgeTypesCoder(idl);
    }
    return TokenBridgeCoder;
}());
exports.TokenBridgeCoder = TokenBridgeCoder;
