// @certusone/wormhole-sdk/lib/esm/solana/wormhole/message.js
import { PublicKey } from "@solana/web3.js";
var MessageData = /** @class */ (function () {
    function MessageData(vaaVersion, consistencyLevel, vaaTime, vaaSignatureAccount, submissionTime, nonce, sequence, emitterChain, emitterAddress, payload) {
        this.vaaVersion = vaaVersion;
        this.consistencyLevel = consistencyLevel;
        this.vaaTime = vaaTime;
        this.vaaSignatureAccount = new PublicKey(vaaSignatureAccount);
        this.submissionTime = submissionTime;
        this.nonce = nonce;
        this.sequence = sequence;
        this.emitterChain = emitterChain;
        this.emitterAddress = emitterAddress;
        this.payload = payload;
    }
    MessageData.deserialize = function (data) {
        var vaaVersion = data.readUInt8(0);
        var consistencyLevel = data.readUInt8(1);
        var vaaTime = data.readUInt32LE(2);
        var vaaSignatureAccount = new PublicKey(data.subarray(6, 38));
        var submissionTime = data.readUInt32LE(38);
        var nonce = data.readUInt32LE(42);
        var sequence = data.readBigUInt64LE(46);
        var emitterChain = data.readUInt16LE(54);
        var emitterAddress = data.subarray(56, 88);
        // unnecessary to get Vec<u8> length, but being explicit in borsh deserialization
        var payloadLen = data.readUInt32LE(88);
        var payload = data.subarray(92, 92 + payloadLen);
        return new MessageData(vaaVersion, consistencyLevel, vaaTime, vaaSignatureAccount, submissionTime, nonce, sequence, emitterChain, emitterAddress, payload);
    };
    return MessageData;
}());
export { MessageData };
