// @certusone/wormhole-sdk/lib/esm/vaa/tokenBridge.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { BN } from "@project-serum/anchor";
import { parseGovernanceVaa } from "./governance";
import { parseVaa } from "./wormhole";
export var TokenBridgePayload;
(function (TokenBridgePayload) {
    TokenBridgePayload[TokenBridgePayload["Transfer"] = 1] = "Transfer";
    TokenBridgePayload[TokenBridgePayload["AttestMeta"] = 2] = "AttestMeta";
    TokenBridgePayload[TokenBridgePayload["TransferWithPayload"] = 3] = "TransferWithPayload";
})(TokenBridgePayload || (TokenBridgePayload = {}));
export var TokenBridgeGovernanceAction;
(function (TokenBridgeGovernanceAction) {
    TokenBridgeGovernanceAction[TokenBridgeGovernanceAction["RegisterChain"] = 1] = "RegisterChain";
    TokenBridgeGovernanceAction[TokenBridgeGovernanceAction["UpgradeContract"] = 2] = "UpgradeContract";
})(TokenBridgeGovernanceAction || (TokenBridgeGovernanceAction = {}));
export function parseTokenTransferPayload(payload) {
    var payloadType = payload.readUInt8(0);
    if (payloadType != TokenBridgePayload.Transfer &&
        payloadType != TokenBridgePayload.TransferWithPayload) {
        throw new Error("not token bridge transfer VAA");
    }
    var amount = BigInt(new BN(payload.subarray(1, 33)).toString());
    var tokenAddress = payload.subarray(33, 65);
    var tokenChain = payload.readUInt16BE(65);
    var to = payload.subarray(67, 99);
    var toChain = payload.readUInt16BE(99);
    var fee = payloadType == 1
        ? BigInt(new BN(payload.subarray(101, 133)).toString())
        : null;
    var fromAddress = payloadType == 3 ? payload.subarray(101, 133) : null;
    var tokenTransferPayload = payload.subarray(133);
    return {
        payloadType: payloadType,
        amount: amount,
        tokenAddress: tokenAddress,
        tokenChain: tokenChain,
        to: to,
        toChain: toChain,
        fee: fee,
        fromAddress: fromAddress,
        tokenTransferPayload: tokenTransferPayload,
    };
}
export function parseTokenTransferVaa(vaa) {
    var parsed = parseVaa(vaa);
    return __assign(__assign({}, parsed), parseTokenTransferPayload(parsed.payload));
}
export function parseAttestMetaPayload(payload) {
    var payloadType = payload.readUInt8(0);
    if (payloadType != TokenBridgePayload.AttestMeta) {
        throw new Error("not token bridge attest meta VAA");
    }
    var tokenAddress = payload.subarray(1, 33);
    var tokenChain = payload.readUInt16BE(33);
    var decimals = payload.readUInt8(35);
    var symbol = payload.subarray(36, 68).toString().replace(/\0/g, "");
    var name = payload.subarray(68, 100).toString().replace(/\0/g, "");
    return {
        payloadType: payloadType,
        tokenAddress: tokenAddress,
        tokenChain: tokenChain,
        decimals: decimals,
        symbol: symbol,
        name: name,
    };
}
export function parseAttestMetaVaa(vaa) {
    var parsed = parseVaa(vaa);
    return __assign(__assign({}, parsed), parseAttestMetaPayload(parsed.payload));
}
export function parseTokenBridgeRegisterChainGovernancePayload(payload) {
    var foreignChain = payload.readUInt16BE(0);
    var foreignAddress = payload.subarray(2, 34);
    return {
        foreignChain: foreignChain,
        foreignAddress: foreignAddress,
    };
}
export function parseTokenBridgeRegisterChainVaa(vaa) {
    var parsed = parseGovernanceVaa(vaa);
    if (parsed.action != TokenBridgeGovernanceAction.RegisterChain) {
        throw new Error("parsed.action != TokenBridgeGovernanceAction.RegisterChain");
    }
    return __assign(__assign({}, parsed), parseTokenBridgeRegisterChainGovernancePayload(parsed.orderPayload));
}
export function parseTokenBridgeUpgradeContractGovernancePayload(payload) {
    var newContract = payload.subarray(0, 32);
    return {
        newContract: newContract,
    };
}
export function parseTokenBridgeUpgradeContractVaa(vaa) {
    var parsed = parseGovernanceVaa(vaa);
    if (parsed.action != TokenBridgeGovernanceAction.UpgradeContract) {
        throw new Error("parsed.action != TokenBridgeGovernanceAction.UpgradeContract");
    }
    return __assign(__assign({}, parsed), parseTokenBridgeUpgradeContractGovernancePayload(parsed.orderPayload));
}
