// @certusone/wormhole-sdk/lib/esm/vaa/wormhole.js
import { keccak256 } from "../utils";
export { isBytes } from "ethers/lib/utils";
export function parseVaa(vaa) {
    var signedVaa = Buffer.isBuffer(vaa) ? vaa : Buffer.from(vaa);
    var sigStart = 6;
    var numSigners = signedVaa[5];
    var sigLength = 66;
    var guardianSignatures = [];
    for (var i = 0; i < numSigners; ++i) {
        var start = sigStart + i * sigLength;
        guardianSignatures.push({
            index: signedVaa[start],
            signature: signedVaa.subarray(start + 1, start + 66),
        });
    }
    var body = signedVaa.subarray(sigStart + sigLength * numSigners);
    return {
        version: signedVaa[0],
        guardianSetIndex: signedVaa.readUInt32BE(1),
        guardianSignatures: guardianSignatures,
        timestamp: body.readUInt32BE(0),
        nonce: body.readUInt32BE(4),
        emitterChain: body.readUInt16BE(8),
        emitterAddress: body.subarray(10, 42),
        sequence: body.readBigUInt64BE(42),
        consistencyLevel: body[50],
        payload: body.subarray(51),
        hash: keccak256(body),
    };
}
