// @acala-network/contracts/typechain/factories/contracts/incentives/InterfaceIncentives__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterfaceIncentives__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "enum InterfaceIncentives.PoolId",
                name: "pool",
                type: "uint8",
            },
            {
                indexed: true,
                internalType: "address",
                name: "poolCurrencyId",
                type: "address",
            },
        ],
        name: "ClaimedRewards",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "DepositedShare",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "WithdrewShare",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "enum InterfaceIncentives.PoolId",
                name: "pool",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "poolCurrencyId",
                type: "address",
            },
        ],
        name: "claimRewards",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "depositDexShare",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "enum InterfaceIncentives.PoolId",
                name: "pool",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "poolCurrencyId",
                type: "address",
            },
        ],
        name: "getClaimRewardDeductionRate",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "enum InterfaceIncentives.PoolId",
                name: "pool",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "poolCurrencyId",
                type: "address",
            },
            {
                internalType: "address",
                name: "rewardCurrencyId",
                type: "address",
            },
        ],
        name: "getIncentiveRewardAmount",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address[]",
                name: "currencyIds",
                type: "address[]",
            },
            {
                internalType: "enum InterfaceIncentives.PoolId",
                name: "pool",
                type: "uint8",
            },
            {
                internalType: "address",
                name: "poolCurrencyId",
                type: "address",
            },
            {
                internalType: "address",
                name: "who",
                type: "address",
            },
        ],
        name: "getPendingRewards",
        outputs: [
            {
                internalType: "uint256[]",
                name: "",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "currencyId",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "withdrawDexShare",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class InterfaceIncentives__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.InterfaceIncentives__factory = InterfaceIncentives__factory;
InterfaceIncentives__factory.abi = _abi;
