// @certusone/wormhole-sdk/lib/cjs/vaa/wormhole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseVaa = exports.isBytes = void 0;
var utils_1 = require("../utils");
var utils_2 = require("ethers/lib/utils");
Object.defineProperty(exports, "isBytes", { enumerable: true, get: function () { return utils_2.isBytes; } });
function parseVaa(vaa) {
    var signedVaa = Buffer.isBuffer(vaa) ? vaa : Buffer.from(vaa);
    var sigStart = 6;
    var numSigners = signedVaa[5];
    var sigLength = 66;
    var guardianSignatures = [];
    for (var i = 0; i < numSigners; ++i) {
        var start = sigStart + i * sigLength;
        guardianSignatures.push({
            index: signedVaa[start],
            signature: signedVaa.subarray(start + 1, start + 66),
        });
    }
    var body = signedVaa.subarray(sigStart + sigLength * numSigners);
    return {
        version: signedVaa[0],
        guardianSetIndex: signedVaa.readUInt32BE(1),
        guardianSignatures: guardianSignatures,
        timestamp: body.readUInt32BE(0),
        nonce: body.readUInt32BE(4),
        emitterChain: body.readUInt16BE(8),
        emitterAddress: body.subarray(10, 42),
        sequence: body.readBigUInt64BE(42),
        consistencyLevel: body[50],
        payload: body.subarray(51),
        hash: utils_1.keccak256(body),
    };
}
exports.parseVaa = parseVaa;
