// @certusone/wormhole-sdk/lib/cjs/solana/nftBridge/accounts/wrapped.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrappedMeta = exports.getWrappedMeta = exports.deriveWrappedMintKey = exports.deriveWrappedMetaKey = void 0;
var anchor_1 = require("@project-serum/anchor");
var utils_1 = require("../../../utils");
var utils_2 = require("../../utils");
var tokenBridge_1 = require("../../tokenBridge");
var tokenBridge_2 = require("../../tokenBridge");
Object.defineProperty(exports, "deriveWrappedMetaKey", { enumerable: true, get: function () { return tokenBridge_2.deriveWrappedMetaKey; } });
function deriveWrappedMintKey(tokenBridgeProgramId, tokenChain, tokenAddress, tokenId) {
    if (tokenChain == utils_1.CHAIN_ID_SOLANA) {
        throw new Error("tokenChain == CHAIN_ID_SOLANA does not have wrapped mint key");
    }
    if (typeof tokenAddress == "string") {
        tokenAddress = utils_1.tryNativeToUint8Array(tokenAddress, tokenChain);
    }
    return utils_2.deriveAddress([
        Buffer.from("wrapped"),
        (function () {
            var buf = Buffer.alloc(2);
            buf.writeUInt16BE(tokenChain);
            return buf;
        })(),
        tokenAddress,
        new anchor_1.BN(tokenId.toString()).toArrayLike(Buffer, "be", 32),
    ], tokenBridgeProgramId);
}
exports.deriveWrappedMintKey = deriveWrappedMintKey;
function getWrappedMeta(connection, tokenBridgeProgramId, mint, commitment) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, connection
                    .getAccountInfo(tokenBridge_1.deriveWrappedMetaKey(tokenBridgeProgramId, mint), commitment)
                    .then(function (info) { return WrappedMeta.deserialize(utils_2.getAccountData(info)); })];
        });
    });
}
exports.getWrappedMeta = getWrappedMeta;
var WrappedMeta = /** @class */ (function () {
    function WrappedMeta(chain, tokenAddress, tokenId) {
        this.chain = chain;
        this.tokenAddress = tokenAddress;
        this.tokenId = tokenId;
    }
    WrappedMeta.deserialize = function (data) {
        if (data.length != 66) {
            throw new Error("data.length != 66");
        }
        var chain = data.readUInt16LE(0);
        var tokenAddress = data.subarray(2, 34);
        var tokenId = BigInt(new anchor_1.BN(data.subarray(34, 66), undefined, "le").toString());
        return new WrappedMeta(chain, tokenAddress, tokenId);
    };
    return WrappedMeta;
}());
exports.WrappedMeta = WrappedMeta;
