// @certusone/wormhole-sdk/lib/esm/solana/wormhole/cpi.js
import { deriveFeeCollectorKey, deriveWormholeBridgeDataKey, getEmitterKeys, } from "./accounts";
import { getPostMessageAccounts } from "./instructions";
/**
 * Generate Wormhole PDAs.
 *
 * @param cpiProgramId
 * @param wormholeProgramId
 * @returns
 */
export function getWormholeDerivedAccounts(cpiProgramId, wormholeProgramId) {
    var _a = getEmitterKeys(cpiProgramId, wormholeProgramId), wormholeEmitter = _a.emitter, wormholeSequence = _a.sequence;
    return {
        wormholeBridge: deriveWormholeBridgeDataKey(wormholeProgramId),
        wormholeEmitter: wormholeEmitter,
        wormholeSequence: wormholeSequence,
        wormholeFeeCollector: deriveFeeCollectorKey(wormholeProgramId),
    };
}
/**
 * Generate accounts needed to perform `post_message` instruction
 * as cross-program invocation.
 *
 * @param cpiProgramId
 * @param wormholeProgramId
 * @param payer
 * @param message
 * @returns
 */
export function getPostMessageCpiAccounts(cpiProgramId, wormholeProgramId, payer, message) {
    var accounts = getPostMessageAccounts(wormholeProgramId, payer, cpiProgramId, message);
    return {
        payer: accounts.payer,
        wormholeBridge: accounts.bridge,
        wormholeMessage: accounts.message,
        wormholeEmitter: accounts.emitter,
        wormholeSequence: accounts.sequence,
        wormholeFeeCollector: accounts.feeCollector,
        clock: accounts.clock,
        rent: accounts.rent,
        systemProgram: accounts.systemProgram,
    };
}
