// @acala-network/contracts/typechain/factories/contracts/liquid-crowdloan/ILiquidCrowdloan__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ILiquidCrowdloan,
  ILiquidCrowdloanInterface,
} from "../../../contracts/liquid-crowdloan/ILiquidCrowdloan";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "sender",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "redeemAmount",
        type: "uint256",
      },
    ],
    name: "Redeemed",
    type: "event",
  },
  {
    inputs: [],
    name: "getRedeemCurrency",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "redeem",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ILiquidCrowdloan__factory {
  static readonly abi = _abi;
  static createInterface(): ILiquidCrowdloanInterface {
    return new utils.Interface(_abi) as ILiquidCrowdloanInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ILiquidCrowdloan {
    return new Contract(address, _abi, signerOrProvider) as ILiquidCrowdloan;
  }
}
