// @acala-network/contracts/typechain/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Xtokens__factory = exports.IXtokens__factory = exports.Token__factory = exports.IToken__factory = exports.StableAsset__factory = exports.IStableAsset__factory = exports.Schedule__factory = exports.ISchedule__factory = exports.Oracle__factory = exports.IOracle__factory = exports.NFT__factory = exports.LiquidCrowdloan__factory = exports.ILiquidCrowdloan__factory = exports.InterfaceIncentives__factory = exports.Incentives__factory = exports.IHonzon__factory = exports.Honzon__factory = exports.IHoma__factory = exports.Homa__factory = exports.IEVM__factory = exports.EVM__factory = exports.IEVMAccounts__factory = exports.EVMAccounts__factory = exports.IDEX__factory = exports.DEX__factory = exports.IERC20__factory = exports.factories = void 0;
exports.factories = __importStar(require("./factories"));
var IERC20__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/IERC20__factory");
Object.defineProperty(exports, "IERC20__factory", { enumerable: true, get: function () { return IERC20__factory_1.IERC20__factory; } });
var DEX__factory_1 = require("./factories/contracts/dex/DEX__factory");
Object.defineProperty(exports, "DEX__factory", { enumerable: true, get: function () { return DEX__factory_1.DEX__factory; } });
var IDEX__factory_1 = require("./factories/contracts/dex/IDEX__factory");
Object.defineProperty(exports, "IDEX__factory", { enumerable: true, get: function () { return IDEX__factory_1.IDEX__factory; } });
var EVMAccounts__factory_1 = require("./factories/contracts/evm-accounts/EVMAccounts__factory");
Object.defineProperty(exports, "EVMAccounts__factory", { enumerable: true, get: function () { return EVMAccounts__factory_1.EVMAccounts__factory; } });
var IEVMAccounts__factory_1 = require("./factories/contracts/evm-accounts/IEVMAccounts__factory");
Object.defineProperty(exports, "IEVMAccounts__factory", { enumerable: true, get: function () { return IEVMAccounts__factory_1.IEVMAccounts__factory; } });
var EVM__factory_1 = require("./factories/contracts/evm/EVM__factory");
Object.defineProperty(exports, "EVM__factory", { enumerable: true, get: function () { return EVM__factory_1.EVM__factory; } });
var IEVM__factory_1 = require("./factories/contracts/evm/IEVM__factory");
Object.defineProperty(exports, "IEVM__factory", { enumerable: true, get: function () { return IEVM__factory_1.IEVM__factory; } });
var Homa__factory_1 = require("./factories/contracts/homa/Homa__factory");
Object.defineProperty(exports, "Homa__factory", { enumerable: true, get: function () { return Homa__factory_1.Homa__factory; } });
var IHoma__factory_1 = require("./factories/contracts/homa/IHoma__factory");
Object.defineProperty(exports, "IHoma__factory", { enumerable: true, get: function () { return IHoma__factory_1.IHoma__factory; } });
var Honzon__factory_1 = require("./factories/contracts/honzon/Honzon__factory");
Object.defineProperty(exports, "Honzon__factory", { enumerable: true, get: function () { return Honzon__factory_1.Honzon__factory; } });
var IHonzon__factory_1 = require("./factories/contracts/honzon/IHonzon__factory");
Object.defineProperty(exports, "IHonzon__factory", { enumerable: true, get: function () { return IHonzon__factory_1.IHonzon__factory; } });
var Incentives__factory_1 = require("./factories/contracts/incentives/Incentives__factory");
Object.defineProperty(exports, "Incentives__factory", { enumerable: true, get: function () { return Incentives__factory_1.Incentives__factory; } });
var InterfaceIncentives__factory_1 = require("./factories/contracts/incentives/InterfaceIncentives__factory");
Object.defineProperty(exports, "InterfaceIncentives__factory", { enumerable: true, get: function () { return InterfaceIncentives__factory_1.InterfaceIncentives__factory; } });
var ILiquidCrowdloan__factory_1 = require("./factories/contracts/liquid-crowdloan/ILiquidCrowdloan__factory");
Object.defineProperty(exports, "ILiquidCrowdloan__factory", { enumerable: true, get: function () { return ILiquidCrowdloan__factory_1.ILiquidCrowdloan__factory; } });
var LiquidCrowdloan__factory_1 = require("./factories/contracts/liquid-crowdloan/LiquidCrowdloan__factory");
Object.defineProperty(exports, "LiquidCrowdloan__factory", { enumerable: true, get: function () { return LiquidCrowdloan__factory_1.LiquidCrowdloan__factory; } });
var NFT__factory_1 = require("./factories/contracts/nft/NFT__factory");
Object.defineProperty(exports, "NFT__factory", { enumerable: true, get: function () { return NFT__factory_1.NFT__factory; } });
var IOracle__factory_1 = require("./factories/contracts/oracle/IOracle__factory");
Object.defineProperty(exports, "IOracle__factory", { enumerable: true, get: function () { return IOracle__factory_1.IOracle__factory; } });
var Oracle__factory_1 = require("./factories/contracts/oracle/Oracle__factory");
Object.defineProperty(exports, "Oracle__factory", { enumerable: true, get: function () { return Oracle__factory_1.Oracle__factory; } });
var ISchedule__factory_1 = require("./factories/contracts/schedule/ISchedule__factory");
Object.defineProperty(exports, "ISchedule__factory", { enumerable: true, get: function () { return ISchedule__factory_1.ISchedule__factory; } });
var Schedule__factory_1 = require("./factories/contracts/schedule/Schedule__factory");
Object.defineProperty(exports, "Schedule__factory", { enumerable: true, get: function () { return Schedule__factory_1.Schedule__factory; } });
var IStableAsset__factory_1 = require("./factories/contracts/stable-asset/IStableAsset__factory");
Object.defineProperty(exports, "IStableAsset__factory", { enumerable: true, get: function () { return IStableAsset__factory_1.IStableAsset__factory; } });
var StableAsset__factory_1 = require("./factories/contracts/stable-asset/StableAsset__factory");
Object.defineProperty(exports, "StableAsset__factory", { enumerable: true, get: function () { return StableAsset__factory_1.StableAsset__factory; } });
var IToken__factory_1 = require("./factories/contracts/token/IToken__factory");
Object.defineProperty(exports, "IToken__factory", { enumerable: true, get: function () { return IToken__factory_1.IToken__factory; } });
var Token__factory_1 = require("./factories/contracts/token/Token__factory");
Object.defineProperty(exports, "Token__factory", { enumerable: true, get: function () { return Token__factory_1.Token__factory; } });
var IXtokens__factory_1 = require("./factories/contracts/xtokens/IXtokens__factory");
Object.defineProperty(exports, "IXtokens__factory", { enumerable: true, get: function () { return IXtokens__factory_1.IXtokens__factory; } });
var Xtokens__factory_1 = require("./factories/contracts/xtokens/Xtokens__factory");
Object.defineProperty(exports, "Xtokens__factory", { enumerable: true, get: function () { return Xtokens__factory_1.Xtokens__factory; } });
