// @acala-network/contracts/typechain/factories/contracts/evm-accounts/IEVMAccounts__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IEVMAccounts__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "accountId",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "evmAddress",
                type: "address",
            },
        ],
        name: "ClaimAccount",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "accountId",
                type: "bytes32",
            },
        ],
        name: "claimDefaultEvmAddress",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "evmAddress",
                type: "address",
            },
        ],
        name: "getAccountId",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "accountId",
                type: "bytes32",
            },
        ],
        name: "getEvmAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IEVMAccounts__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IEVMAccounts__factory = IEVMAccounts__factory;
IEVMAccounts__factory.abi = _abi;
