// @acala-network/contracts/typechain/factories/contracts/schedule/ISchedule__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ISchedule__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "taskId",
                type: "bytes",
            },
        ],
        name: "CanceledCall",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "taskId",
                type: "bytes",
            },
        ],
        name: "RescheduledCall",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "contractAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "taskId",
                type: "bytes",
            },
        ],
        name: "ScheduledCall",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "taskId",
                type: "bytes",
            },
        ],
        name: "cancelCall",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "minDelay",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "taskId",
                type: "bytes",
            },
        ],
        name: "rescheduleCall",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "contractAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "value",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "storageLimit",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "minDelay",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "inputData",
                type: "bytes",
            },
        ],
        name: "scheduleCall",
        outputs: [
            {
                internalType: "bytes",
                name: "",
                type: "bytes",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ISchedule__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ISchedule__factory = ISchedule__factory;
ISchedule__factory.abi = _abi;
